data_steward/analytics/cdr_ops/combined.py

```diff
@@ -0,0 +1,479 @@
+# values in this dataset should generally be non-increasing compared to baseline.
+# If the prevalence of invalid concepts is increasing, EHR Ops should be alerted.
+
+FQN_INVALID_CONCEPT = 'curation_sandbox.invalid_concept'
+
+query = f'''
+CREATE OR REPLACE TABLE `{FQN_INVALID_CONCEPT}` 
+(dataset_id  STRING NOT NULL OPTIONS(description="Identifies the dataset")
+,table_name  STRING NOT NULL OPTIONS(description="Identifies the table")
+,column_name STRING NOT NULL OPTIONS(description="Identifies the concept_id field")
+,src_hpo_id  STRING NOT NULL OPTIONS(description="Identifies the HPO that supplied the records")
+,concept_id  INT64           OPTIONS(description="A value that is not in the associated concept table (can be NULL)")
+,row_count   INT64  NOT NULL)
+OPTIONS
+(description='Count of invalid values provided in required concept_id fields for multiple datasets.'
+,expiration_timestamp=TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL 7 DAY) )
+'''
+q(query)
+
+tpl = JINJA_ENV.from_string('''
+-- Construct a query that finds all rows in all tables --
+-- where concept_id fields have an invalid value and reports --
+-- the row count per site --
+DECLARE query DEFAULT (
+    SELECT 
+ STRING_AGG('SELECT ' 
+     || '"{{DATASET_ID}}"'  || ' AS dataset_id '
+     || ',"' || table_name  || '"  AS table_name '
+     || ',"' || column_name || '" AS column_name '
+     || ', m.src_hpo_id AS src_hpo_id '
+     || ',' || column_name  || ' AS concept_id '
+     || ', COUNT(1) AS row_count '
+     -- domain table joined with its mapping table --
+     || 'FROM `' || table_schema || '.' || table_name || '` t'
+     || ' JOIN `' || table_schema || '.' || '_mapping_' || table_name || '` m'
+     || '  USING (' || c.table_name || '_id) '
+     || ' LEFT JOIN `{{DATASET_ID}}.concept` c '
+     || '  ON ' || column_name || ' = c.concept_id '
+     
+     -- invalid concept_id --
+     || 'WHERE c.concept_id IS NULL '
+     || 'GROUP BY 2, 3, 4, 5',
+           ' UNION ALL ')
+FROM `{{DATASET_ID}}.INFORMATION_SCHEMA.COLUMNS` c
+JOIN `{{DATASET_ID}}.__TABLES__` t
+ ON c.table_name = t.table_id
+WHERE 1=1
+  -- the table has data --
+  AND t.row_count > 0
+  -- there is an associated mapping table --
+  AND EXISTS (
+       SELECT 1 
+       FROM `{{DATASET_ID}}.INFORMATION_SCHEMA.COLUMNS` 
+      WHERE table_name = '_mapping_' || c.table_name
+        AND column_name = c.table_name || '_id'
+  )
+  -- the column is a required concept_id --
+  AND NOT ENDS_WITH(LOWER(column_name), 'source_concept_id')
+  AND ENDS_WITH(LOWER(column_name), 'concept_id')
+  AND is_nullable = 'NO'
+);
+EXECUTE IMMEDIATE (
+ 'INSERT `{{FQN_INVALID_CONCEPT}}` (dataset_id, table_name, column_name, src_hpo_id, concept_id, row_count) ' || query);
+''')
+q(tpl.render(DATASET_ID=DATASET_ID, FQN_INVALID_CONCEPT=FQN_INVALID_CONCEPT))
+q(
+    tpl.render(DATASET_ID=BASELINE_DATASET_ID,
+               FQN_INVALID_CONCEPT=FQN_INVALID_CONCEPT))
+
+q(f'''
```

Comment:
This is another instance of using `q()` directly instead of defining the string and using it.  It's an optional fix, but seems to go against the code style in the rest of this script.